sub-packages/bionemo-llm/src/bionemo/llm/lightning.py

```diff
@@ -305,11 +308,19 @@ def forward_step(self, batch) -> Tensor:
 
     def training_step(self, batch, batch_idx: Optional[int] = None) -> Tensor:
         """In mcore the loss-function is part of the forward-pass when labels are provided."""
-        return self.forward_step(batch)
+        outputs = self.forward_step(batch)
+        logits = outputs["token_logits"].transpose(0, 1)  #  [s, b] -> [b, s]
+        self.train_ppl(logits, batch["labels"])
+        self.log("train_ppl_step", self.train_ppl)
+        return outputs
 
     def validation_step(self, batch, batch_idx: Optional[int] = None) -> Tensor:
         """In mcore the loss-function is part of the forward-pass when labels are provided."""
-        return self.forward_step(batch)
+        outputs = self.forward_step(batch)
+        logits = outputs["token_logits"].transpose(0, 1)  #  [s, b] -> [b, s]
+        self.valid_ppl(logits, batch["labels"])
+        self.log("valid_ppl_step", self.valid_ppl)
```

Comment:
But we will need `sync_dist=True` and `on_epoch=True` here.

Reply:
why is that? It seems to give similar perplexity values to the callback:
https://wandb.ai/clara-discovery/bionemo-hf-pretraining/reports/val_ppl-valid_ppl_step-24-11-20-18-12-31---VmlldzoxMDI2NjQ0MA

Reply:
> The sync_dist, sync_dist_group and reduce_fx flags from self.log(...) don’t affect the metric logging in any manner. The metric class contains its own distributed synchronization logic.
https://lightning.ai/docs/torchmetrics/stable/pages/lightning.html#logging-torchmetrics

So I think we _do_ want `on_epoch=True` here (I'm not sure why my example run was so close without it...), but I think sync_dist is handled internally?

Reply:
You need `sync_dist_group` because you don't want to sync with 0/other weird value from non-last-pp-stage devices.

Reply:
Maybe they are the same on single device because you had `self.log("valid_ppl_step", self.valid_ppl)` in `on_valid_epoch_end`?